### ISSUE TITLE
Override custom variant with theme

### ISSUE THREAD

**Issue by ludovicmnji** [styled-system/styled-system#706]
I'm using the utility `variant()` to define some generic styles (with styled-system v5.1.0). I would like to be able to redefine some of the variants inside my custom theme but without having to re-define all the variants. Is it possible to do that ?

This is a short example of what I want to achieve :

```js
// Heading.jsx

import styled from 'styled-components';
import { variant, color, space } from 'styled-system';

const variants = variant({
  scale: 'titles',
  variants: {
    title1: {
      fontSize: 7,
      lineHeight: 1.5,
    },
    title2: {
      fontSize: 6,
      lineHeight: 1.5,
    },
    title3: {
      fontSize: 5,
      lineHeight: 1.5,
    },
    title4: {
      fontSize: 4,
      lineHeight: 1.2,
    },
    title5: {
      fontSize: 3,
      lineHeight: 1.1,
    },
    title6: {
      fontSize: 2,
      lineHeight: 1.1,
      textTransform: 'uppercase',
    },
  }
});

const Heading = styled.h1`
  font-family: ${({ theme }) => theme.fonts.primary};
  ${variants}
  ${color}
  ${space}
`;

export default Heading;
```

```js
// theme/index.js

export default {
// ...
  titles: {
    title1: {
      fontSize: 7,
      lineHeight: 1.5,
      textTransform: 'uppercase',
    }
  },
  // ...
}
```

If I do like that, it doesn't work as I expected. I lose the rest of variants (title2, title3, title4 ...) and I have to redefine inside my theme.

**Comment by jxnblk** [comment]
Currently it does not merge the inline and theme-based objects, but I could see this being a sensible addition to the API, see https://github.com/styled-system/styled-system/blob/master/packages/core/src/index.js#L48

**Comment by ludovicmnji** [comment]
Thanks for the answer, I was not sure to use `variant` correctly or if it was just a bug. 

I don't know the complexity for you to implement that, but in my opinion it could be very powerfull. It could participate in writing more customizable component with an abstract way (across theme). Immediately, for my use, I'm going to write base style inside my default theme whereas component definition and rewrite each style in others themes.
  
Do you have an idea for planning this improvement ?

**Comment by jxnblk** [comment]
Yeah, I don't imagine it'll be too difficult to implement, but also need to be cautious about breaking changes and performance implications (which is the more time consuming part). Going to leave this issue open for now. Thanks!

**Comment by harrisrobin** [comment]
@jxnblk thanks for leaving this open, we also ran into this issue and were wondering if it was a bug or intended. 

I think this would be a great addition to the API, let me know if there is anything I can do to help get this enhancement implemented :)

**Comment by maxparelius** [comment]
I agree with @harrisrobin it seems limiting to not be able to use a variant and also manually override one the styles with a prop.

**Comment by mathanpec** [comment]
@jxnblk: Is it still open? Can i take stab at this?

**Comment by jxnblk** [comment]
@mathanpec sure thing! Let me know if you have any questions about it

**Comment by mathanpec** [comment]
@jxnblk: Have raised a PR #886 for this. Please take a look at it.

**Comment by mattstyles** [comment]
In lieu of the PR to do this at library level you can do it in userland by exposing the variants object and merging yourself in the theme.

```js
const variants = {
  primary: {
    bg: 'blue',
    color: 'white'
  },
  secondary: {
    bg: 'papayawhip',
    color: 'black'
  }
}

const colourVariants = variant({
  scale: 'colours',
  variants: variants
})

const theme = {
   // ....
  colours: {
    ...variants,
    outrun: {
      bg: 'indigo',
      color: 'magenta'
    }
  }
}
```

This has the advantage of being able to extend the core variants you expose from your library, with the con that its a bit more typing and scaffold for consumers.

**Comment by thiagonunesbatista** [comment]
I was receiving the "breakpoints.map is not a function" error when I was trying to use the variant from Styled-System. The approach of @mattstyles worked for me. I receive the props inside the styled component, then I pass the value to an external function that return an object of styles, I consider this an overengineering, and this will not integrate with the theme, but for what I need it's working.

![Screenshot from 2020-12-06 12-33-50](https://user-images.githubusercontent.com/67487679/101284587-8b488900-37bf-11eb-90d9-9322cff7b5de.png)

![Screenshot from 2020-12-06 12-35-57](https://user-images.githubusercontent.com/67487679/101284597-9d2a2c00-37bf-11eb-9b0b-2fdf250b3bb2.png)
